### DISCUSSION TITLE
Testing private rxMethods in SignalStore

### DISCUSSION THREAD

**Discussion by MotassemJa** [ngrx/platform#4523] [Q&A] [answered]
I have a signalStore that has a private method to fetch some data and I call it in the `onInit` Hook of the store:
```typescript
withMethods(store => ({
  _fetchData: rxMethod<void>(pipe(
    switchMap(() => inject(Service).getData().pipe(
      ...
    ))
  ))
})),
withHooks({
  onInit(store) {
    store._fetchData()
  }
})
````

How can I unit test the private method with jest?

**Comment by twittwer** [comment]
Do you want to test the internals or the behaviour of the store?

For the store behaviour, you may spy on the service method, instantiate the store and check for the service call.

**Reply by MotassemJa** [reply]
I want to test the logic when the `_fetchData` is called. Like when an error is thrown by the service for example, the error state should be updated.

**Reply by rainerhahnekamp** [reply, marked as answer]
Every private method has an impact on the outside/public. In your case, you want to test that the error state is updated.

Provide a test double for `Service`, let it throw an error, and verify that the state got the expected change.

**Reply by MotassemJa** [reply]
Okay sounds fair. But for some reason it seems that the `onInit` Hook is not being called when configuring the test with TestBed:
```typescript
const getDataMock = jest.fn().mockImplementation(() => of([dataStub()]));
beforeEach(() => {
  TestBed.configureTestingModule({
    providers: [
      MySignalStore, 
      { provide: Service, useValue: { getData: getDataMock }}
    ]
  });
  store = TestBed.inject(MySignalStore);
});

it('should call getData', () => {
  expect(getDataMock).toHaveBeenCalled() // fails
})
```

**Reply by MotassemJa** [reply]
I had a typo in my providers array which caused the error, sorry for that.
